src/handles/tpm.rs

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    constants::tss::{
+        TPM2_HT_AC, TPM2_HT_HMAC_SESSION, TPM2_HT_NV_INDEX, TPM2_HT_PCR, TPM2_HT_PERMANENT,
+        TPM2_HT_PERSISTENT, TPM2_HT_POLICY_SESSION, TPM2_HT_TRANSIENT,
+    },
+    tss2_esys::TPM2_HANDLE,
+    Error, Result, WrapperErrorKind,
+};
+use log::error;
+use std::convert::{From, TryFrom};
+/// Struct representing the most general form
+/// of a TPM handle.
+///
+/// * Details
+/// The TPM handles are used
+/// to reference shielded locations of various
+/// types within the TPM.
+///
+/// * OBS
+/// Do not confuse the TpmHandles with the
+/// ESYS [ObjectHandle](crate::handles::esys::ObjectHandle).
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct TpmHandle {
+    value: u32,
+}
+
+impl TpmHandle {
+    pub fn create_from_value(value: u32) -> Result<TpmHandle> {
+        let most_significant_byte = value.to_be_bytes()[0];
+        match most_significant_byte {
+            TPM2_HT_PCR => Ok(TpmHandle { value }),            // 0x00
+            TPM2_HT_NV_INDEX => Ok(TpmHandle { value }),       // 0x01
+            TPM2_HT_HMAC_SESSION => Ok(TpmHandle { value }), // 0x02, Same as TPM2_HT_LOADED_SESSION
+            TPM2_HT_POLICY_SESSION => Ok(TpmHandle { value }), // 0x03, Same as TPM2_HT_SAVED_SESSION
+            TPM2_HT_PERMANENT => Ok(TpmHandle { value }),      // 0x04
+            TPM2_HT_TRANSIENT => Ok(TpmHandle { value }),      // 0x80
+            TPM2_HT_PERSISTENT => Ok(TpmHandle { value }),     // 0x81
+            TPM2_HT_AC => Ok(TpmHandle { value }),             // 0x90
```

Comment:
If the TPM handles are part of a set, maybe it would be nice for the `TpmHandle` type to be an enum of those different categories. Maybe that will help later with static checks.

Reply:
My thinking is that TpmHandle is the base and all specific version of TpmHandle can always be converted to TpmHandle. 

So if we take the NvIndex type from the NV storage PR(#104)

I would rename NvIndex to NvIndexTpmHandle and move that type into src/handles/tpm.rs
```rust
pub struct NvIndexTpmHandle {
    tpm_handle: TpmHandle,
}

impl NvIndexTpmHandle {
   pub fn new(value: u32) -> Result<NvIndexTpmHandle> {
       if value.to_be_bytes()[0] != TPM2_HT_NV_INDEX {
            error!("Error: Invalid most significant byte is invalid (!={})", TPM2_HT_NV_INDEX);
            Err....
       }
       NvIndexTpmHandle { value: TpmHandle {value} }
    }
}

imp From<NvIndexTpmHandle> for TpmHandle {
    fn from(nv_index_tpm_handle: NvIndexTpmHandle) -> TpmHandle {
         nv_index_tpm_handle.value    
    }
}
```

Then the functions that only uses nv index tpm handles can clearly state that and any attempts at trying to use the wrong type of handle will be catched at compile time.

I am note sure I understood what it was you suggested so could you please elaborate?

Reply:
> Then the functions that only uses nv index tpm handles can clearly state that and any attempts at trying to use the wrong type of handle will be catched at compile time.

Agree with that and what you said!
I was thinknig taking it a little step further and having:
```rust
pub enum TpmHandle {
    Pcr(Pcr),
    NvIndex(NvIndex),
    ... etc
}

pub struct NvIndex {
    value: u32,
}

imp From<NvIndex> for TpmHandle {
    fn from(nv_index: NvIndex) -> TpmHandle {
         TpmHandle::NvIndex(nv_index)    
    }
}
```
We use this idiom quite a lot in other repos and I think it's nice 😃 But agree with yours as well.

Reply:
Hrrm. Yeah I agree with you. It will be easier to do it your way for APIs that can return any type of TPM handle. I changed it. I also fixed that none of the new types were publicly available (hehe oops).